{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0fac22ce-be26-459c-9c62-45c4325da539",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import subprocess\n",
    "from collections import namedtuple\n",
    "\n",
    "# external modules \n",
    "import requests\n",
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "612dd80b-a9e9-4c5e-929a-be1e653fc4f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "multifasta_proteins = '../../results/cassettes/clusters/merged_cassette_proteins.faa.gz'\n",
    "clusters = '../../results/cassettes/clusters/mmseqs2_cluster.tsv'\n",
    "annotation_dir = '../../results/Prokka/'\n",
    "defense_system_prediction_dir = '../../results/cassettes/extracted/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a1e3580b-f321-4d99-8b93-84a24c80b0a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def cassettes_headers_to_df(multifasta_proteins):\n",
    "    cmd = ['zgrep', '-Po', '(?<=^\\>).+', multifasta_proteins]\n",
    "    columns = ['contigs', 'cassettes', 'locus_tags', 'headers']\n",
    "    data = []\n",
    "    for header in subprocess.run(cmd, check=True, stdout=subprocess.PIPE).stdout.decode('UTF-8').splitlines():\n",
    "        rows = header.split(':')\n",
    "        assert len(rows) == 3, 'Header {} is invalid !'.format(header)\n",
    "        rows.append(header)\n",
    "        data.append(rows)\n",
    "        \n",
    "    df = pd.DataFrame(data=data, columns=columns) \\\n",
    "            .assign(genome_ids = lambda x: x['contigs'].str.split(':') \\\n",
    "                .apply(lambda x: '_'.join(x[0].split('_')[1:-1])))\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6f613bde-5c23-4ed2-834b-ba44be9cd4cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "headers_df = cassettes_headers_to_df(multifasta_proteins=multifasta_proteins)\n",
    "clusters_df = pd.read_csv(clusters, sep='\\t', header=None).rename(columns={0:'representatives', 1:'members'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1199415a-7c29-435e-9d59-1ec340c69cee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "representatives\n",
       "Contig_1395103.4_47:Cassette_1:KDHGNBDG_03652    14.0\n",
       "Contig_936157.3_23:Cassette_1:IOMDNFNB_01075     12.0\n",
       "Contig_936157.3_23:Cassette_1:IOMDNFNB_01077     12.0\n",
       "Contig_936157.3_62:Cassette_1:IOMDNFNB_04497     12.0\n",
       "Contig_936157.3_62:Cassette_1:IOMDNFNB_04496     12.0\n",
       "                                                 ... \n",
       "Contig_1395103.4_5:Cassette_1:KDHGNBDG_00493      2.0\n",
       "Contig_28901.3694_7:Cassette_5:HDGJAOOK_01579     2.0\n",
       "Contig_28901.4009_1:Cassette_2:BKDLJELO_00369     2.0\n",
       "Contig_28901.3694_7:Cassette_5:HDGJAOOK_01568     2.0\n",
       "Contig_28901.4009_1:Cassette_2:BKDLJELO_00371     2.0\n",
       "Name: representatives, Length: 237, dtype: float64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count mmseqs clusters with more than one sequence (botton length value is count)\n",
    "clusters_df.groupby('representatives')['representatives'].count().sort_values(ascending=False).where(lambda x: x > 1).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b90bf173-9327-4693-aa13-97f0036ec55a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contigs</th>\n",
       "      <th>cassettes</th>\n",
       "      <th>locus_tags</th>\n",
       "      <th>headers</th>\n",
       "      <th>genome_ids</th>\n",
       "      <th>is_rep</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Contig_936157.3_6</td>\n",
       "      <td>Cassette_1</td>\n",
       "      <td>IOMDNFNB_00323</td>\n",
       "      <td>Contig_936157.3_6:Cassette_1:IOMDNFNB_00323</td>\n",
       "      <td>936157.3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Contig_936157.3_6</td>\n",
       "      <td>Cassette_1</td>\n",
       "      <td>IOMDNFNB_00324</td>\n",
       "      <td>Contig_936157.3_6:Cassette_1:IOMDNFNB_00324</td>\n",
       "      <td>936157.3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Contig_936157.3_6</td>\n",
       "      <td>Cassette_1</td>\n",
       "      <td>IOMDNFNB_00325</td>\n",
       "      <td>Contig_936157.3_6:Cassette_1:IOMDNFNB_00325</td>\n",
       "      <td>936157.3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Contig_936157.3_6</td>\n",
       "      <td>Cassette_1</td>\n",
       "      <td>IOMDNFNB_00326</td>\n",
       "      <td>Contig_936157.3_6:Cassette_1:IOMDNFNB_00326</td>\n",
       "      <td>936157.3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Contig_936157.3_6</td>\n",
       "      <td>Cassette_1</td>\n",
       "      <td>IOMDNFNB_00328</td>\n",
       "      <td>Contig_936157.3_6:Cassette_1:IOMDNFNB_00328</td>\n",
       "      <td>936157.3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             contigs   cassettes      locus_tags  \\\n",
       "0  Contig_936157.3_6  Cassette_1  IOMDNFNB_00323   \n",
       "1  Contig_936157.3_6  Cassette_1  IOMDNFNB_00324   \n",
       "2  Contig_936157.3_6  Cassette_1  IOMDNFNB_00325   \n",
       "3  Contig_936157.3_6  Cassette_1  IOMDNFNB_00326   \n",
       "5  Contig_936157.3_6  Cassette_1  IOMDNFNB_00328   \n",
       "\n",
       "                                       headers genome_ids  is_rep  \n",
       "0  Contig_936157.3_6:Cassette_1:IOMDNFNB_00323   936157.3    True  \n",
       "1  Contig_936157.3_6:Cassette_1:IOMDNFNB_00324   936157.3    True  \n",
       "2  Contig_936157.3_6:Cassette_1:IOMDNFNB_00325   936157.3    True  \n",
       "3  Contig_936157.3_6:Cassette_1:IOMDNFNB_00326   936157.3    True  \n",
       "5  Contig_936157.3_6:Cassette_1:IOMDNFNB_00328   936157.3    True  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if sequence is a mmseqs clusters representative\n",
    "headers_df = headers_df.assign(is_rep = lambda df: list(map(lambda header: header in clusters_df.representatives.values, df.headers.values)))\n",
    "headers_df[headers_df.is_rep].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b1f92eb0-723a-40ac-bba1-495633e3cd37",
   "metadata": {},
   "outputs": [],
   "source": [
    "def list_annotations_files(annotation_dir: str) -> dict:\n",
    "    cmd = [\n",
    "        'find',\n",
    "        annotation_dir,\n",
    "        '-type', 'f',\n",
    "        '-name', '*.gbk'\n",
    "    ]\n",
    "    files = {}\n",
    "    for file in subprocess.run(cmd, check=True, stdout=subprocess.PIPE).stdout.decode('UTF-8').splitlines():\n",
    "        file = Path(file)\n",
    "        genome_id = file.stem.replace('_out', '')\n",
    "        genome_path = str(file.absolute())\n",
    "        assert not files.get(genome_id), 'Repeated ids: {}'.format(genome_id)\n",
    "        files[genome_id] = genome_path\n",
    "        \n",
    "    return files\n",
    "\n",
    "def retrive_features_from_gbk(df, gbk_path):\n",
    "    my_feature = namedtuple('Feature', 'locus_tag gene product translation')\n",
    "    wanted_locus_tags = set(df.locus_tags.values)\n",
    "    CDSs = []\n",
    "    # Read files and get the features of the representative clusters\n",
    "    with open(gbk_path) as f:\n",
    "        for record in SeqIO.parse(gbk_path, 'genbank'):\n",
    "            for feature in filter(lambda fet: fet.type == 'CDS' and fet.qualifiers.get('locus_tag', [None])[0] in wanted_locus_tags, record.features):\n",
    "                get_qual = lambda x: feature.qualifiers.get(x, [None])[0]\n",
    "                locus_tag = get_qual('locus_tag')\n",
    "                gene = get_qual('gene')\n",
    "                gene = 'Unk' if not gene else gene\n",
    "                product = get_qual('product')\n",
    "                product = 'Unk' if product == 'hypothetical protein' else product \n",
    "                translation = get_qual('translation')\n",
    "                CDSs.append((\n",
    "                    locus_tag,\n",
    "                    my_feature(\n",
    "                        locus_tag,\n",
    "                        gene,\n",
    "                        product,\n",
    "                        translation,\n",
    "                    )\n",
    "                ))\n",
    "    \n",
    "    CDSs = dict(CDSs)\n",
    "    # Add features to dataframe\n",
    "    return df.assign(features = lambda df: df.locus_tags.apply(lambda locus_tag: CDSs.get(locus_tag, None)))\n",
    "                \n",
    "                \n",
    "def locus_tag_to_edges(topology: list, drop_duplicates=False):\n",
    "    edges = []\n",
    "    for indx, cluster_id in enumerate(topology):\n",
    "        if indx == 0:\n",
    "            edges.append((cluster_id, topology[indx+1]))\n",
    "        elif indx == len(topology) - 1:\n",
    "            edges.append((cluster_id, topology[indx-1]))\n",
    "        else:\n",
    "            edges.append((cluster_id, topology[indx-1]))\n",
    "            edges.append((cluster_id, topology[indx+1]))\n",
    "    if not drop_duplicates:\n",
    "        return edges\n",
    "    \n",
    "    strict_edges = []\n",
    "    for edge in edges:\n",
    "        if not ( edge in strict_edges or edge[::-1] in strict_edges ):\n",
    "            strict_edges.append(edge)\n",
    "    return strict_edges     \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "64e79ade-454a-48d3-8d90-ea76cdcb790b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contigs</th>\n",
       "      <th>cassettes</th>\n",
       "      <th>locus_tags</th>\n",
       "      <th>headers</th>\n",
       "      <th>genome_ids</th>\n",
       "      <th>is_rep</th>\n",
       "      <th>features</th>\n",
       "      <th>cluster_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Contig_1395103.4_5</td>\n",
       "      <td>Cassette_1</td>\n",
       "      <td>KDHGNBDG_00483</td>\n",
       "      <td>Contig_1395103.4_5:Cassette_1:KDHGNBDG_00483</td>\n",
       "      <td>1395103.4</td>\n",
       "      <td>True</td>\n",
       "      <td>(KDHGNBDG_00483, rep_2, ATP-dependent DNA heli...</td>\n",
       "      <td>ID_9_rep_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Contig_1395103.4_5</td>\n",
       "      <td>Cassette_1</td>\n",
       "      <td>KDHGNBDG_00487</td>\n",
       "      <td>Contig_1395103.4_5:Cassette_1:KDHGNBDG_00487</td>\n",
       "      <td>1395103.4</td>\n",
       "      <td>True</td>\n",
       "      <td>(KDHGNBDG_00487, proA_1, Gamma-glutamyl phosph...</td>\n",
       "      <td>ID_11_proA_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Contig_1395103.4_5</td>\n",
       "      <td>Cassette_1</td>\n",
       "      <td>KDHGNBDG_00488</td>\n",
       "      <td>Contig_1395103.4_5:Cassette_1:KDHGNBDG_00488</td>\n",
       "      <td>1395103.4</td>\n",
       "      <td>True</td>\n",
       "      <td>(KDHGNBDG_00488, proB, Glutamate 5-kinase, MSD...</td>\n",
       "      <td>ID_12_proB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Contig_1395103.4_5</td>\n",
       "      <td>Cassette_1</td>\n",
       "      <td>KDHGNBDG_00490</td>\n",
       "      <td>Contig_1395103.4_5:Cassette_1:KDHGNBDG_00490</td>\n",
       "      <td>1395103.4</td>\n",
       "      <td>True</td>\n",
       "      <td>(KDHGNBDG_00490, crl, Sigma factor-binding pro...</td>\n",
       "      <td>ID_13_crl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Contig_1395103.4_5</td>\n",
       "      <td>Cassette_1</td>\n",
       "      <td>KDHGNBDG_00492</td>\n",
       "      <td>Contig_1395103.4_5:Cassette_1:KDHGNBDG_00492</td>\n",
       "      <td>1395103.4</td>\n",
       "      <td>True</td>\n",
       "      <td>(KDHGNBDG_00492, gpt, Xanthine phosphoribosylt...</td>\n",
       "      <td>ID_15_gpt</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               contigs   cassettes      locus_tags  \\\n",
       "9   Contig_1395103.4_5  Cassette_1  KDHGNBDG_00483   \n",
       "11  Contig_1395103.4_5  Cassette_1  KDHGNBDG_00487   \n",
       "12  Contig_1395103.4_5  Cassette_1  KDHGNBDG_00488   \n",
       "13  Contig_1395103.4_5  Cassette_1  KDHGNBDG_00490   \n",
       "15  Contig_1395103.4_5  Cassette_1  KDHGNBDG_00492   \n",
       "\n",
       "                                         headers genome_ids  is_rep  \\\n",
       "9   Contig_1395103.4_5:Cassette_1:KDHGNBDG_00483  1395103.4    True   \n",
       "11  Contig_1395103.4_5:Cassette_1:KDHGNBDG_00487  1395103.4    True   \n",
       "12  Contig_1395103.4_5:Cassette_1:KDHGNBDG_00488  1395103.4    True   \n",
       "13  Contig_1395103.4_5:Cassette_1:KDHGNBDG_00490  1395103.4    True   \n",
       "15  Contig_1395103.4_5:Cassette_1:KDHGNBDG_00492  1395103.4    True   \n",
       "\n",
       "                                             features   cluster_ids  \n",
       "9   (KDHGNBDG_00483, rep_2, ATP-dependent DNA heli...    ID_9_rep_2  \n",
       "11  (KDHGNBDG_00487, proA_1, Gamma-glutamyl phosph...  ID_11_proA_1  \n",
       "12  (KDHGNBDG_00488, proB, Glutamate 5-kinase, MSD...    ID_12_proB  \n",
       "13  (KDHGNBDG_00490, crl, Sigma factor-binding pro...     ID_13_crl  \n",
       "15  (KDHGNBDG_00492, gpt, Xanthine phosphoribosylt...     ID_15_gpt  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Retrive feature (gene name, translation, locus_tag) information of clusters representatives \n",
    "annotation_files = list_annotations_files(annotation_dir)\n",
    "\n",
    "rep_with_features = []\n",
    "for genome, df in headers_df[headers_df.is_rep].groupby('genome_ids'):\n",
    "    gbk_path = annotation_files.get(genome)\n",
    "    if gbk_path:\n",
    "        rep_with_features.append(retrive_features_from_gbk(df, gbk_path))\n",
    "    \n",
    "get_clusters_ids = lambda df: list(map(lambda x: 'ID_{}_{}'.format(*x),zip(list(map(str, df.index.values)), list(map(lambda fet: fet.gene, df.features)))))\n",
    "rep_with_features_df = pd.concat(rep_with_features).reset_index(drop=True).assign(cluster_ids = lambda df: get_clusters_ids(df))\n",
    "rep_with_features_df[~rep_with_features_df.cluster_ids.str.contains('Unk')].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "c010ebb8-2ee6-4131-9e9c-5a7cc66c6ed0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get nodes and edges\n",
    "neighborhood_gdf = clusters_df \\\n",
    "        .merge(right=rep_with_features_df.loc[:,['headers', 'cluster_ids']], left_on='representatives', right_on='headers') \\\n",
    "        .drop(['representatives', 'headers'], axis=1) \\\n",
    "        .assign(cassettes = lambda df: tuple(map(lambda member: ':'.join(member.split(':')[:2]), df.members))) \\\n",
    "        .assign(locus_tags = lambda df: tuple(map(lambda member: member.split(':')[2], df.members))) \\\n",
    "        .groupby('cassettes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "76a0cced-b0c0-4581-b0d1-9ece41070e72",
   "metadata": {},
   "outputs": [],
   "source": [
    "edges = []\n",
    "for cassette, df in neighborhood_gdf:\n",
    "    # make shore that the locus_tags are ordered otherwise we lost\n",
    "    # topology relationships.\n",
    "    \n",
    "    df = df.sort_values(['locus_tags'], key=lambda locus_tags: list(map(lambda x: int(x[1]), locus_tags.str.split('_').values)))\n",
    "    cassette_edges = pd.DataFrame(data=locus_tag_to_edges(topology=df.loc[:,['cluster_ids']].cluster_ids.values, drop_duplicates=True), columns=['source', 'target'])\n",
    "    \n",
    "    # edges.append(pd.merge(\n",
    "    #     left=cassette_edges,\n",
    "    #     right=rep_with_features_df.loc[:,['cluster_ids', 'features']].assign(features = lambda df: list(map(lambda fet: str(fet).replace('Feature(', '').replace(')', '').replace(',',';').replace(' ',''), df.features))),\n",
    "    #     left_on='sources',\n",
    "    #     right_on='cluster_ids',\n",
    "    #     how='inner'\n",
    "    # ).drop('cluster_ids', axis=1).rename(columns={'features':'source_features'}))\n",
    "    \n",
    "    edges.append(cassette_edges)\n",
    "    \n",
    "\n",
    "edges_df = pd.concat(edges).drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "72e183a2-dcf1-4cc7-8970-3a442ff2d890",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading Javascript client ... 855b6e9c-dc35-451e-842a-e055ba8ff0df on https://jupyter-bridge.cytoscape.org\n",
      "You are connected to Cytoscape!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'apiVersion': 'v1',\n",
       " 'cytoscapeVersion': '3.9.1',\n",
       " 'automationAPIVersion': '1.6.0',\n",
       " 'py4cytoscapeVersion': '1.5.0'}"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import requests\n",
    "exec(requests.get(\"https://raw.githubusercontent.com/cytoscape/jupyter-bridge/master/client/p4c_init.py\").text)\n",
    "IPython.display.Javascript(_PY4CYTOSCAPE_BROWSER_CLIENT_JS) # Start browser client\n",
    "import py4cytoscape as p4c\n",
    "p4c.cytoscape_ping()\n",
    "p4c.cytoscape_version_info()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "87b7bea6-f039-45ad-8883-3723172eefab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# nodes_df = edges_df.loc[:,['source']].rename(columns={'source':'id'})\n",
    "# p4c.create_network_from_data_frames(nodes_df, edges_df, title=\"Salmonella Defense Systems\", collection=\"DataFrame Example\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "284eb4f2-aa36-4f64-ab4d-0f70c9df310b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p4c.delete_all_networks()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "39be5010-cade-46e9-bd90-a0b96ff06093",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tags</th>\n",
       "      <th>names</th>\n",
       "      <th>systems</th>\n",
       "      <th>tool</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BKDLJELO_00330</td>\n",
       "      <td>MTase_I</td>\n",
       "      <td>DMS_other</td>\n",
       "      <td>padloc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BKDLJELO_00330</td>\n",
       "      <td>MTase_I</td>\n",
       "      <td>RM_type_I</td>\n",
       "      <td>padloc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BKDLJELO_00331</td>\n",
       "      <td>Specificity_I</td>\n",
       "      <td>DMS_other</td>\n",
       "      <td>padloc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BKDLJELO_00331</td>\n",
       "      <td>Specificity_I</td>\n",
       "      <td>RM_type_I</td>\n",
       "      <td>padloc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BKDLJELO_00332</td>\n",
       "      <td>REase_I</td>\n",
       "      <td>DMS_other</td>\n",
       "      <td>padloc</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       locus_tags          names    systems    tool\n",
       "0  BKDLJELO_00330        MTase_I  DMS_other  padloc\n",
       "1  BKDLJELO_00330        MTase_I  RM_type_I  padloc\n",
       "2  BKDLJELO_00331  Specificity_I  DMS_other  padloc\n",
       "3  BKDLJELO_00331  Specificity_I  RM_type_I  padloc\n",
       "4  BKDLJELO_00332        REase_I  DMS_other  padloc"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cmd = ['find', defense_system_prediction_dir, '-name', '*merged_defense_systems_prediction.csv']\n",
    "file_paths = subprocess.run(cmd, check=True, stdout=subprocess.PIPE).stdout.decode('UTF-8').splitlines()\n",
    "defense_genes_prediction_df = pd.concat(map(lambda path: pd.read_csv(path, sep=','), file_paths))\n",
    "defense_genes_prediction_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "b2eb40d7-7748-401d-9e48-bf648e4a63fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tags</th>\n",
       "      <th>predictions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>IOMDNFNB_04284</td>\n",
       "      <td>tool=padloc,systems=druantia_type_II,names=DruK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>IOMDNFNB_04490</td>\n",
       "      <td>tool=padloc,systems=DMS_other,names=Specificit...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>IOMDNFNB_04491</td>\n",
       "      <td>tool=padloc,systems=DMS_other,names=MTase_I;to...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>IOMDNFNB_04492</td>\n",
       "      <td>tool=padloc,systems=DMS_other,names=REase_I;to...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>IOMDNFNB_04493</td>\n",
       "      <td>tool=padloc,systems=DMS_other,names=mREase_IV;...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        locus_tags                                        predictions\n",
       "44  IOMDNFNB_04284    tool=padloc,systems=druantia_type_II,names=DruK\n",
       "45  IOMDNFNB_04490  tool=padloc,systems=DMS_other,names=Specificit...\n",
       "46  IOMDNFNB_04491  tool=padloc,systems=DMS_other,names=MTase_I;to...\n",
       "47  IOMDNFNB_04492  tool=padloc,systems=DMS_other,names=REase_I;to...\n",
       "48  IOMDNFNB_04493  tool=padloc,systems=DMS_other,names=mREase_IV;..."
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "format_values = lambda x: '='.join(x)\n",
    "\n",
    "map_defence_prediction = []\n",
    "for locus_tag, df in defense_genes_prediction_df.head(100).groupby('locus_tags'):\n",
    "    df = df.fillna('None')\n",
    "    prediction_string = ';'.join(list(map(lambda x: ','.join(x),df.drop('locus_tags', axis=1).apply(lambda x: x .name + '=' + x).iloc[:,::-1].values)))\n",
    "    map_defence_prediction.append((locus_tag, prediction_string))\n",
    "\n",
    "map_defence_prediction_df = pd.DataFrame(map_defence_prediction, columns=['locus_tags', 'predictions']).drop_duplicates()\n",
    "map_defence_prediction_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "3cfac263-7a8d-431b-9f12-724c07f55f96",
   "metadata": {},
   "outputs": [],
   "source": [
    "oi = map_defence_prediction_df.merge(\n",
    "    headers_df.loc[:,['locus_tags', 'headers']],\n",
    "    on='locus_tags',\n",
    "    how='inner'\n",
    ").groupby('locus_tags')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "dfaba84a-84eb-4302-8a54-cade3d8c778f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        locus_tags                                        predictions  \\\n",
      "26  IOMDNFNB_01071  tool=padloc,systems=DMS_other,names=oMTase;too...   \n",
      "27  IOMDNFNB_01071  tool=padloc,systems=DMS_other,names=oMTase;too...   \n",
      "\n",
      "                                         headers  \n",
      "26  Contig_936157.3_23:Cassette_1:IOMDNFNB_01071  \n",
      "27  Contig_936157.3_23:Cassette_2:IOMDNFNB_01071  \n",
      "        locus_tags                                        predictions  \\\n",
      "28  IOMDNFNB_01072  tool=padloc,systems=DMS_other,names=REase_II;t...   \n",
      "29  IOMDNFNB_01072  tool=padloc,systems=DMS_other,names=REase_II;t...   \n",
      "\n",
      "                                         headers  \n",
      "28  Contig_936157.3_23:Cassette_1:IOMDNFNB_01072  \n",
      "29  Contig_936157.3_23:Cassette_2:IOMDNFNB_01072  \n",
      "        locus_tags                                        predictions  \\\n",
      "30  IOMDNFNB_02069  tool=padloc,systems=druantia_type_III,names=DruH3   \n",
      "31  IOMDNFNB_02069  tool=padloc,systems=druantia_type_III,names=DruH3   \n",
      "\n",
      "                                         headers  \n",
      "30  Contig_936157.3_37:Cassette_1:IOMDNFNB_02069  \n",
      "31  Contig_936157.3_37:Cassette_2:IOMDNFNB_02069  \n",
      "        locus_tags                                        predictions  \\\n",
      "32  IOMDNFNB_02070  tool=padloc,systems=druantia_type_III,names=DruE3   \n",
      "33  IOMDNFNB_02070  tool=padloc,systems=druantia_type_III,names=DruE3   \n",
      "\n",
      "                                         headers  \n",
      "32  Contig_936157.3_37:Cassette_1:IOMDNFNB_02070  \n",
      "33  Contig_936157.3_37:Cassette_2:IOMDNFNB_02070  \n",
      "        locus_tags                                        predictions  \\\n",
      "42  IOMDNFNB_03816  tool=padloc,systems=DRT_other,names=NDT_II-A;t...   \n",
      "43  IOMDNFNB_03816  tool=padloc,systems=DRT_other,names=NDT_II-A;t...   \n",
      "\n",
      "                                         headers  \n",
      "42  Contig_936157.3_54:Cassette_1:IOMDNFNB_03816  \n",
      "43  Contig_936157.3_54:Cassette_2:IOMDNFNB_03816  \n",
      "        locus_tags                                        predictions  \\\n",
      "44  IOMDNFNB_03817  tool=padloc,systems=DRT_other,names=RT_II-A;to...   \n",
      "45  IOMDNFNB_03817  tool=padloc,systems=DRT_other,names=RT_II-A;to...   \n",
      "\n",
      "                                         headers  \n",
      "44  Contig_936157.3_54:Cassette_1:IOMDNFNB_03817  \n",
      "45  Contig_936157.3_54:Cassette_2:IOMDNFNB_03817  \n",
      "        locus_tags                                        predictions  \\\n",
      "51  IOMDNFNB_04490  tool=padloc,systems=DMS_other,names=Specificit...   \n",
      "52  IOMDNFNB_04490  tool=padloc,systems=DMS_other,names=Specificit...   \n",
      "53  IOMDNFNB_04490  tool=padloc,systems=DMS_other,names=Specificit...   \n",
      "\n",
      "                                         headers  \n",
      "51  Contig_936157.3_62:Cassette_1:IOMDNFNB_04490  \n",
      "52  Contig_936157.3_62:Cassette_2:IOMDNFNB_04490  \n",
      "53  Contig_936157.3_62:Cassette_3:IOMDNFNB_04490  \n",
      "        locus_tags                                        predictions  \\\n",
      "54  IOMDNFNB_04491  tool=padloc,systems=DMS_other,names=MTase_I;to...   \n",
      "55  IOMDNFNB_04491  tool=padloc,systems=DMS_other,names=MTase_I;to...   \n",
      "56  IOMDNFNB_04491  tool=padloc,systems=DMS_other,names=MTase_I;to...   \n",
      "\n",
      "                                         headers  \n",
      "54  Contig_936157.3_62:Cassette_1:IOMDNFNB_04491  \n",
      "55  Contig_936157.3_62:Cassette_2:IOMDNFNB_04491  \n",
      "56  Contig_936157.3_62:Cassette_3:IOMDNFNB_04491  \n",
      "        locus_tags                                        predictions  \\\n",
      "57  IOMDNFNB_04492  tool=padloc,systems=DMS_other,names=REase_I;to...   \n",
      "58  IOMDNFNB_04492  tool=padloc,systems=DMS_other,names=REase_I;to...   \n",
      "59  IOMDNFNB_04492  tool=padloc,systems=DMS_other,names=REase_I;to...   \n",
      "\n",
      "                                         headers  \n",
      "57  Contig_936157.3_62:Cassette_1:IOMDNFNB_04492  \n",
      "58  Contig_936157.3_62:Cassette_2:IOMDNFNB_04492  \n",
      "59  Contig_936157.3_62:Cassette_3:IOMDNFNB_04492  \n",
      "        locus_tags                                        predictions  \\\n",
      "60  IOMDNFNB_04493  tool=padloc,systems=DMS_other,names=mREase_IV;...   \n",
      "61  IOMDNFNB_04493  tool=padloc,systems=DMS_other,names=mREase_IV;...   \n",
      "62  IOMDNFNB_04493  tool=padloc,systems=DMS_other,names=mREase_IV;...   \n",
      "\n",
      "                                         headers  \n",
      "60  Contig_936157.3_62:Cassette_1:IOMDNFNB_04493  \n",
      "61  Contig_936157.3_62:Cassette_2:IOMDNFNB_04493  \n",
      "62  Contig_936157.3_62:Cassette_3:IOMDNFNB_04493  \n"
     ]
    }
   ],
   "source": [
    "for i, df in oi:\n",
    "    if df.shape[0] > 1:\n",
    "        print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aade4162-add2-4bff-9369-b18adac396e2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
