{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0fac22ce-be26-459c-9c62-45c4325da539",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import subprocess\n",
    "from collections import namedtuple\n",
    "from collections import defaultdict\n",
    "import gzip\n",
    "from functools import reduce\n",
    "from functools import partial\n",
    "import re\n",
    "import itertools\n",
    "\n",
    "\n",
    "# external modules \n",
    "import requests\n",
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a1e3580b-f321-4d99-8b93-84a24c80b0a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def cassettes_headers_to_df(multifasta_proteins):\n",
    "    cmd = ['zgrep', '-Po', '(?<=^\\>).+', multifasta_proteins]\n",
    "    columns = ['contigs', 'cassettes', 'locus_tags', 'headers']\n",
    "    data = []\n",
    "    for header in subprocess.run(cmd, check=True, stdout=subprocess.PIPE).stdout.decode('UTF-8').splitlines():\n",
    "        rows = header.split(':')\n",
    "        assert len(rows) == 3, 'Header {} is invalid !'.format(header)\n",
    "        rows.append(header)\n",
    "        data.append(rows)\n",
    "        \n",
    "    df = pd.DataFrame(data=data, columns=columns) \\\n",
    "            .assign(genome_ids = lambda x: x['contigs'].str.split(':') \\\n",
    "                .apply(lambda x: '_'.join(x[0].split('_')[1:-1])))\n",
    "    return df\n",
    "\n",
    "\n",
    "                \n",
    "def get_edges(topology: list, drop_duplicates=True):\n",
    "    edges = []\n",
    "    for indx, cluster_id in enumerate(topology):\n",
    "        if indx == 0:\n",
    "            edges.append((cluster_id, topology[indx+1]))\n",
    "        elif indx == len(topology) - 1:\n",
    "            edges.append((cluster_id, topology[indx-1]))\n",
    "        else:\n",
    "            edges.append((cluster_id, topology[indx-1]))\n",
    "            edges.append((cluster_id, topology[indx+1]))\n",
    "    if not drop_duplicates:\n",
    "        return edges\n",
    "    \n",
    "    strict_edges = []\n",
    "    for edge in edges:\n",
    "        if not ( edge in strict_edges or edge[::-1] in strict_edges ):\n",
    "            strict_edges.append(edge)\n",
    "    return strict_edges\n",
    "\n",
    "def get_wanted_locus_tags_from_gbk(gbk_path: str, wanted_locus_tags: list):\n",
    "    with open(gbk_path) as f:\n",
    "        features = []\n",
    "        for record in SeqIO.parse(gbk_path, 'genbank'):\n",
    "            for feature in filter(lambda fet: fet.qualifiers.get('locus_tag', [None])[0] in wanted_locus_tags, record.features):\n",
    "                features.append((feature.qualifiers.get('locus_tag', [None])[0], feature))\n",
    "        return pd.DataFrame(data=features, columns=['locus_tags', 'features'])\n",
    "    \n",
    "def list_gbk_files(annotation_dir: str) -> dict:\n",
    "    cmd = [\n",
    "        'find',\n",
    "        annotation_dir,\n",
    "        '-type', 'f',\n",
    "        '-name', '*.gbk'\n",
    "    ]\n",
    "    files = {}\n",
    "    for file in subprocess.run(cmd, check=True, stdout=subprocess.PIPE).stdout.decode('UTF-8').splitlines():\n",
    "        file = Path(file)\n",
    "        genome_id = file.stem.replace('_out', '')\n",
    "        genome_path = str(file.absolute())\n",
    "        assert not files.get(genome_id), 'Repeated ids: {}'.format(genome_id)\n",
    "        files[genome_id] = genome_path\n",
    "        \n",
    "    return files\n",
    "\n",
    "def get_feature_df(genome_filters: list, annotation_dir: str):\n",
    "    \"\"\"genome_filter = { genome_id : locus_tags }\"\"\"\n",
    "    genome_paths: dict = list_gbk_files(annotation_dir)\n",
    "    \n",
    "    features_df = pd.DataFrame()\n",
    "    for genome_id, locus_tags in genome_filters.items():\n",
    "        df = get_wanted_locus_tags_from_gbk(gbk_path=genome_paths[genome_id], wanted_locus_tags=locus_tags)\n",
    "        if features_df.empty:\n",
    "            features_df = df\n",
    "            continue\n",
    "        features_df = pd.concat([features_df, df])\n",
    "        \n",
    "    return features_df.reset_index(drop=True)\n",
    "        \n",
    "            \n",
    "            \n",
    "def get_feature_string(feature):\n",
    "    get_qualifier = lambda query: feature.qualifiers.get(query, ['NA'])[0]\n",
    "    fields = ('gene', 'locus_tag', 'product', 'translation')\n",
    "    return ';'.join(list(map(lambda field: '='.join((field, get_qualifier(field))) , fields)))\n",
    "\n",
    "\n",
    "def read_and_merge_predictions(defense_system_prediction_dir: str):\n",
    "    cmd = [\n",
    "        'find',\n",
    "        defense_system_prediction_dir,\n",
    "        '-type', 'f',\n",
    "        '-name', '*merged_defense_systems_prediction.csv',\n",
    "        '!', '-empty'\n",
    "    ]\n",
    "    assert Path(defense_system_prediction_dir).exists()\n",
    "    prediction_files: list = subprocess.run(cmd, check=True, stdout=subprocess.PIPE).stdout.decode('UTF-8').splitlines()\n",
    "    assert prediction_files, 'No predict file found !'\n",
    "    read_dataframe = lambda file_path: pd.read_csv(file_path, sep=',')\n",
    "    \n",
    "    return pd.concat(map(read_dataframe, prediction_files)).reset_index(drop=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60c81c97-0c0a-457e-8715-7c43d0f34bc5",
   "metadata": {},
   "source": [
    "# 16K genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "28f5c8d3-4d5d-4e62-abd8-cf46b983f696",
   "metadata": {},
   "outputs": [],
   "source": [
    "tsv_rep = '/home/hugo/projects/salmonella-bacterial-immunity/net_data_16K/mmseqs_cluster.tsv.gz'\n",
    "all_gffs = '/home/hugo/projects/salmonella-bacterial-immunity/net_data_16K/all_csds.gff.gz'\n",
    "all_GOIs = '/home/hugo/projects/salmonella-bacterial-immunity/net_data_16K/prodigal_with_GOIs.csv.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a9f1568c-bd62-4a37-906a-62e56b6e4b0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_rep_locus_tag</th>\n",
       "      <th>contigs</th>\n",
       "      <th>cassettes</th>\n",
       "      <th>is_GOI</th>\n",
       "      <th>strand</th>\n",
       "      <th>locus_tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>JGKINGID_02763</td>\n",
       "      <td>28901.7257_7</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>+</td>\n",
       "      <td>JGKINGID_02763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JGKINGID_02763</td>\n",
       "      <td>90370.2301_37</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>-</td>\n",
       "      <td>JADHFFOP_02760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>JGKINGID_02763</td>\n",
       "      <td>590.14785_3</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>+</td>\n",
       "      <td>HMFKIHPF_02332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JGKINGID_02763</td>\n",
       "      <td>85569.336_1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>+</td>\n",
       "      <td>AEHNPGNC_00428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>JGKINGID_02763</td>\n",
       "      <td>590.16494_2</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>+</td>\n",
       "      <td>IGMGEBKI_01366</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cluster_rep_locus_tag        contigs cassettes is_GOI strand       locus_tag\n",
       "0        JGKINGID_02763   28901.7257_7         1  False      +  JGKINGID_02763\n",
       "1        JGKINGID_02763  90370.2301_37         1  False      -  JADHFFOP_02760\n",
       "2        JGKINGID_02763    590.14785_3         1  False      +  HMFKIHPF_02332\n",
       "3        JGKINGID_02763    85569.336_1         1  False      +  AEHNPGNC_00428\n",
       "4        JGKINGID_02763    590.16494_2         1  False      +  IGMGEBKI_01366"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def tsv_rep_to_df(tsv_rep: str):\n",
    "    strings_to_remove = ('Contig_', 'Cassette_', 'GOI_', 'Strand_', )\n",
    "    with gzip.open(tsv_rep) as f:\n",
    "        rep = None\n",
    "        for line in iter(map(lambda string: string.decode().strip(), f)):\n",
    "            cluster, member = line.split('\\t')\n",
    "            cluster =  cluster.split(':')[-1]\n",
    "            if rep != cluster:\n",
    "                rep = cluster\n",
    "            yield (\n",
    "                cluster,\n",
    "                *reduce(lambda rec, substr: rec.replace(substr, ''), strings_to_remove, member).split(':'),\n",
    "            )\n",
    "\n",
    "clusters_df = pd.DataFrame(tsv_rep_to_df(tsv_rep=tsv_rep), columns=('cluster_rep_locus_tag', 'contigs', 'cassettes', 'is_GOI', 'strand', 'locus_tag'))\n",
    "clusters_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "5c7f3a1b-b2b2-4ed0-af8c-f33540f2d235",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>feature</th>\n",
       "      <th>locus_tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>28901.7257_4</td>\n",
       "      <td>124482</td>\n",
       "      <td>125993</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=JGKINGID_01837;eC_number=3.4.11.1;Name=pepA...</td>\n",
       "      <td>JGKINGID_01837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>28901.7257_4</td>\n",
       "      <td>127488</td>\n",
       "      <td>128570</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=JGKINGID_01839;Name=lptG;db_xref=COG:COG079...</td>\n",
       "      <td>JGKINGID_01839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>28901.7257_4</td>\n",
       "      <td>128765</td>\n",
       "      <td>129763</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=JGKINGID_01840;Name=gntR_3;gene=gntR_3;infe...</td>\n",
       "      <td>JGKINGID_01840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>28901.7257_4</td>\n",
       "      <td>133803</td>\n",
       "      <td>134822</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=JGKINGID_01845;eC_number=1.1.1.2;Name=ahr;d...</td>\n",
       "      <td>JGKINGID_01845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>28901.7257_4</td>\n",
       "      <td>145406</td>\n",
       "      <td>146218</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=JGKINGID_01852;inference=ab initio predicti...</td>\n",
       "      <td>JGKINGID_01852</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      contig_id   start     end strand  \\\n",
       "0  28901.7257_4  124482  125993      -   \n",
       "1  28901.7257_4  127488  128570      +   \n",
       "2  28901.7257_4  128765  129763      -   \n",
       "3  28901.7257_4  133803  134822      -   \n",
       "4  28901.7257_4  145406  146218      -   \n",
       "\n",
       "                                             feature       locus_tag  \n",
       "0  ID=JGKINGID_01837;eC_number=3.4.11.1;Name=pepA...  JGKINGID_01837  \n",
       "1  ID=JGKINGID_01839;Name=lptG;db_xref=COG:COG079...  JGKINGID_01839  \n",
       "2  ID=JGKINGID_01840;Name=gntR_3;gene=gntR_3;infe...  JGKINGID_01840  \n",
       "3  ID=JGKINGID_01845;eC_number=1.1.1.2;Name=ahr;d...  JGKINGID_01845  \n",
       "4  ID=JGKINGID_01852;inference=ab initio predicti...  JGKINGID_01852  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rep_prokka_annotation = pd.concat(\n",
    "    map(lambda df:\n",
    "        (df.iloc[:, [0,3,4,6,8]]\n",
    "         .rename(columns=dict(zip((0,3,4,6,8), ('contig_id', 'start', 'end', 'strand', 'feature'))))\n",
    "         .assign(locus_tag = lambda df: df.feature.apply(lambda feature: dict(map(lambda x: x.split('='), feature.split(';'))).get('locus_tag')))\n",
    "         .query('locus_tag.isin(@clusters_df.cluster_rep_locus_tag)')),\n",
    "        pd.read_csv(all_gffs, chunksize=1000, iterator=True, sep='\\t', header=None))\n",
    ")\n",
    "rep_prokka_annotation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "292afa78-9d89-4a9e-8111-d90b4068dd4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prediction</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>coords</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NA;defensefinder:name=Olokun_OloA,system=Olokun</td>\n",
       "      <td>28901.7369_60</td>\n",
       "      <td>(+, 7164, 9602)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NA;defensefinder:name=Olokun_OloB,system=Olokun</td>\n",
       "      <td>28901.7369_60</td>\n",
       "      <td>(+, 9553, 11025)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>padloc:name=Dsr1,system=dsr1;NA</td>\n",
       "      <td>28901.7369_60</td>\n",
       "      <td>(-, 11996, 15700)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>padloc:name=mREase_IV,system=RM_type_IV;defens...</td>\n",
       "      <td>28901.7369_97</td>\n",
       "      <td>(+, 5128, 5961)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>padloc:name=RT_II-A,system=DRT_other;defensefi...</td>\n",
       "      <td>28901.7369_179</td>\n",
       "      <td>(-, 1625, 2557)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          prediction       contig_id  \\\n",
       "0    NA;defensefinder:name=Olokun_OloA,system=Olokun   28901.7369_60   \n",
       "1    NA;defensefinder:name=Olokun_OloB,system=Olokun   28901.7369_60   \n",
       "2                    padloc:name=Dsr1,system=dsr1;NA   28901.7369_60   \n",
       "3  padloc:name=mREase_IV,system=RM_type_IV;defens...   28901.7369_97   \n",
       "4  padloc:name=RT_II-A,system=DRT_other;defensefi...  28901.7369_179   \n",
       "\n",
       "              coords  \n",
       "0    (+, 7164, 9602)  \n",
       "1   (+, 9553, 11025)  \n",
       "2  (-, 11996, 15700)  \n",
       "3    (+, 5128, 5961)  \n",
       "4    (-, 1625, 2557)  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gcoord = namedtuple('Gcoord', 'strand start end')\n",
    "is_overlapping = lambda a, b: ( a.strand == b.strand ) and ( max(a.start, b.start) < min(a.end, b.end) )\n",
    "\n",
    "prodigal_df = pd.concat(map(\n",
    "    lambda df:\n",
    "        (df.loc[:, ['locus_tag', 'start', 'end', 'strand', 'prediction']]\n",
    "         .assign(strand = lambda df: df.strand.apply(lambda x: {-1 : '-', 1 : '+'}.get(x)))\n",
    "         .assign(contig_id = lambda df: df.locus_tag.apply(lambda x: '_'.join(x.split('_')[:-1])))\n",
    "         .query('contig_id.isin(@rep_prokka_annotation.contig_id)')\n",
    "         .drop('locus_tag', axis=1)\n",
    "         .assign(coords = lambda df: tuple(map(lambda x: gcoord(*x), df.loc[:,['strand', 'start', 'end']].values.tolist())))\n",
    "         .iloc[:, [-3,-2,-1]]),\n",
    "    pd.read_csv(all_GOIs, chunksize=1000, iterator=True, sep=';')\n",
    ")).reset_index(drop=True)\n",
    "        \n",
    "prodigal_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "66381a71-65c7-411d-9ddd-a42846a53ff8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>feature</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>dfs_prediction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>28901.7257_4</td>\n",
       "      <td>124482</td>\n",
       "      <td>125993</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=JGKINGID_01837;eC_number=3.4.11.1;Name=pepA...</td>\n",
       "      <td>JGKINGID_01837</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>28901.7257_4</td>\n",
       "      <td>127488</td>\n",
       "      <td>128570</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=JGKINGID_01839;Name=lptG;db_xref=COG:COG079...</td>\n",
       "      <td>JGKINGID_01839</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>28901.7257_4</td>\n",
       "      <td>128765</td>\n",
       "      <td>129763</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=JGKINGID_01840;Name=gntR_3;gene=gntR_3;infe...</td>\n",
       "      <td>JGKINGID_01840</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>28901.7257_4</td>\n",
       "      <td>133803</td>\n",
       "      <td>134822</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=JGKINGID_01845;eC_number=1.1.1.2;Name=ahr;d...</td>\n",
       "      <td>JGKINGID_01845</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>28901.7257_4</td>\n",
       "      <td>145406</td>\n",
       "      <td>146218</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=JGKINGID_01852;inference=ab initio predicti...</td>\n",
       "      <td>JGKINGID_01852</td>\n",
       "      <td>NA;defensefinder:name=DUF4435,system=Rst_PARIS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      contig_id   start     end strand  \\\n",
       "0  28901.7257_4  124482  125993      -   \n",
       "1  28901.7257_4  127488  128570      +   \n",
       "2  28901.7257_4  128765  129763      -   \n",
       "3  28901.7257_4  133803  134822      -   \n",
       "4  28901.7257_4  145406  146218      -   \n",
       "\n",
       "                                             feature       locus_tag  \\\n",
       "0  ID=JGKINGID_01837;eC_number=3.4.11.1;Name=pepA...  JGKINGID_01837   \n",
       "1  ID=JGKINGID_01839;Name=lptG;db_xref=COG:COG079...  JGKINGID_01839   \n",
       "2  ID=JGKINGID_01840;Name=gntR_3;gene=gntR_3;infe...  JGKINGID_01840   \n",
       "3  ID=JGKINGID_01845;eC_number=1.1.1.2;Name=ahr;d...  JGKINGID_01845   \n",
       "4  ID=JGKINGID_01852;inference=ab initio predicti...  JGKINGID_01852   \n",
       "\n",
       "                                   dfs_prediction  \n",
       "0                                              NA  \n",
       "1                                              NA  \n",
       "2                                              NA  \n",
       "3                                              NA  \n",
       "4  NA;defensefinder:name=DUF4435,system=Rst_PARIS  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_overlapping(a, coords):\n",
    "    is_overlapping = lambda a, b: ( a.strand == b.strand ) and ( max(a.start, b.start) < min(a.end, b.end) )\n",
    "    prediction = coords.get(a)\n",
    "    \n",
    "    if prediction:\n",
    "        return prediction\n",
    "    \n",
    "    for b, prediction in coords.items():\n",
    "        if is_overlapping(a, b):\n",
    "            return prediction\n",
    "        \n",
    "    return 'NA'\n",
    "\n",
    "rep_prokka_annotation_with_dfs_prediction_df = (rep_prokka_annotation\n",
    "     .assign(coords = lambda df: tuple(map(lambda x: gcoord(*x), df.loc[:,['strand', 'start', 'end']].values.tolist())))\n",
    "     .groupby('contig_id', as_index=False)\n",
    "     .pipe(lambda df: df.apply(lambda df: df.assign( dfs_prediction = df.coords.apply(partial(\n",
    "         get_overlapping,\n",
    "         coords=prodigal_df.query('contig_id == @df.contig_id.iloc[0]').loc[:,['prediction', 'coords']].set_index('coords', drop=True).prediction.to_dict())\n",
    "        ))))\n",
    "     .drop('coords', axis=1))\n",
    "\n",
    "rep_prokka_annotation_with_dfs_prediction_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "027e4719-03bd-49e3-a0c4-f9a21818d970",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>edge_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EJAOJOGH_03687</td>\n",
       "      <td>JGKINGID_03495</td>\n",
       "      <td>16118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAADGEIO_02721</td>\n",
       "      <td>LHMLBOEB_03555</td>\n",
       "      <td>16095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HBDDAOCK_04993</td>\n",
       "      <td>JGKINGID_03495</td>\n",
       "      <td>16067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EHEOBCKO_00724</td>\n",
       "      <td>JCMOLLNE_04069</td>\n",
       "      <td>16038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EHEOBCKO_00724</td>\n",
       "      <td>MJIHJCFH_03921</td>\n",
       "      <td>15920</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source          target  edge_count\n",
       "0  EJAOJOGH_03687  JGKINGID_03495       16118\n",
       "1  CAADGEIO_02721  LHMLBOEB_03555       16095\n",
       "2  HBDDAOCK_04993  JGKINGID_03495       16067\n",
       "3  EHEOBCKO_00724  JCMOLLNE_04069       16038\n",
       "4  EHEOBCKO_00724  MJIHJCFH_03921       15920"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_edges = []\n",
    "for cassette, df in clusters_df.groupby(['contigs', 'cassettes'], as_index=False):\n",
    "    try:\n",
    "        cassette_locus_tags: list = (\n",
    "            df.sort_values('locus_tag', key=lambda locus_tag: locus_tag.apply(lambda locus_tag: int(locus_tag.split('_')[1])))\n",
    "            .cluster_rep_locus_tag\n",
    "            .values\n",
    "            .tolist()\n",
    "        )\n",
    "\n",
    "        edges = get_edges(topology=cassette_locus_tags)\n",
    "        edges = iter(map(lambda tup: sorted(tup), edges))\n",
    "        all_edges.append(pd.DataFrame(data=edges, columns=['source', 'target']))\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "edges_df = (pd.concat(all_edges)\n",
    "            .groupby(['source', 'target'], as_index=False)\n",
    "            .size()\n",
    "            .sort_values('size', ascending=False)\n",
    "            .rename(columns={'size' : 'edge_count'})\n",
    "            .reset_index(drop=True))\n",
    "\n",
    "edges_df.to_csv('edges_cache.csv', index=False, sep=';')\n",
    "edges_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "ed030a3f-0113-4315-8948-fe6e0be65f2b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>members_locus_tags</th>\n",
       "      <th>members_count</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>feature</th>\n",
       "      <th>dfs_prediction</th>\n",
       "      <th>most_frq_strnd_orint</th>\n",
       "      <th>most_frq_strnd_orint_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LHMLBOEB_03555</td>\n",
       "      <td>LHMLBOEB_03555,JGKINGID_03493,JADHFFOP_00300,H...</td>\n",
       "      <td>16184</td>\n",
       "      <td>58097.9_19</td>\n",
       "      <td>19746</td>\n",
       "      <td>21710</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=LHMLBOEB_03555;inference=ab initio predicti...</td>\n",
       "      <td>padloc:name=MTase_III,system=DMS_other;defense...</td>\n",
       "      <td>-</td>\n",
       "      <td>8530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EJAOJOGH_03687</td>\n",
       "      <td>EJAOJOGH_03687,JGKINGID_03496,JADHFFOP_00304,H...</td>\n",
       "      <td>16168</td>\n",
       "      <td>1925258.5_15</td>\n",
       "      <td>20928</td>\n",
       "      <td>22202</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=EJAOJOGH_03687;eC_number=7.1.1.3;Name=appC_...</td>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>8522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAADGEIO_02721</td>\n",
       "      <td>CAADGEIO_02721,AGNHIDCH_00165,JGKINGID_03492,J...</td>\n",
       "      <td>16167</td>\n",
       "      <td>90371.894_58</td>\n",
       "      <td>12759</td>\n",
       "      <td>14021</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=CAADGEIO_02721;Name=yjhB_1;gene=yjhB_1;infe...</td>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>8526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JGKINGID_03495</td>\n",
       "      <td>JGKINGID_03495,JADHFFOP_00303,HMFKIHPF_01786,A...</td>\n",
       "      <td>16163</td>\n",
       "      <td>28901.7257_10</td>\n",
       "      <td>155306</td>\n",
       "      <td>155407</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=JGKINGID_03495;inference=ab initio predicti...</td>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>8525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HBDDAOCK_04993</td>\n",
       "      <td>HBDDAOCK_04993,AKAHHPCI_01866,JGKINGID_03494,H...</td>\n",
       "      <td>16117</td>\n",
       "      <td>90371.2142_302</td>\n",
       "      <td>83282</td>\n",
       "      <td>86080</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=HBDDAOCK_04993;inference=ab initio predicti...</td>\n",
       "      <td>NA;defensefinder:name=Type_III_REases,system=RM</td>\n",
       "      <td>-</td>\n",
       "      <td>8504</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               id                                 members_locus_tags  \\\n",
       "0  LHMLBOEB_03555  LHMLBOEB_03555,JGKINGID_03493,JADHFFOP_00300,H...   \n",
       "1  EJAOJOGH_03687  EJAOJOGH_03687,JGKINGID_03496,JADHFFOP_00304,H...   \n",
       "2  CAADGEIO_02721  CAADGEIO_02721,AGNHIDCH_00165,JGKINGID_03492,J...   \n",
       "3  JGKINGID_03495  JGKINGID_03495,JADHFFOP_00303,HMFKIHPF_01786,A...   \n",
       "4  HBDDAOCK_04993  HBDDAOCK_04993,AKAHHPCI_01866,JGKINGID_03494,H...   \n",
       "\n",
       "   members_count       contig_id   start     end strand  \\\n",
       "0          16184      58097.9_19   19746   21710      -   \n",
       "1          16168    1925258.5_15   20928   22202      +   \n",
       "2          16167    90371.894_58   12759   14021      -   \n",
       "3          16163   28901.7257_10  155306  155407      +   \n",
       "4          16117  90371.2142_302   83282   86080      +   \n",
       "\n",
       "                                             feature  \\\n",
       "0  ID=LHMLBOEB_03555;inference=ab initio predicti...   \n",
       "1  ID=EJAOJOGH_03687;eC_number=7.1.1.3;Name=appC_...   \n",
       "2  ID=CAADGEIO_02721;Name=yjhB_1;gene=yjhB_1;infe...   \n",
       "3  ID=JGKINGID_03495;inference=ab initio predicti...   \n",
       "4  ID=HBDDAOCK_04993;inference=ab initio predicti...   \n",
       "\n",
       "                                      dfs_prediction most_frq_strnd_orint  \\\n",
       "0  padloc:name=MTase_III,system=DMS_other;defense...                    -   \n",
       "1                                                 NA                    -   \n",
       "2                                                 NA                    -   \n",
       "3                                                 NA                    -   \n",
       "4    NA;defensefinder:name=Type_III_REases,system=RM                    -   \n",
       "\n",
       "   most_frq_strnd_orint_count  \n",
       "0                        8530  \n",
       "1                        8522  \n",
       "2                        8526  \n",
       "3                        8525  \n",
       "4                        8504  "
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes_in_edges = set(itertools.chain.from_iterable(edges_df.loc[:,[\"source\", \"target\"]].values))\n",
    "\n",
    "nodes = (clusters_df\n",
    "         .groupby('cluster_rep_locus_tag', as_index=False)\n",
    "         .apply(lambda df: pd.DataFrame(\n",
    "             [(df.cluster_rep_locus_tag.values[0], ','.join(df.locus_tag.values.tolist()), len(df.locus_tag.values))],\n",
    "             columns=['cluster_rep_locus_tag', 'members_locus_tags', 'members_count']\n",
    "         ))\n",
    "         .sort_values(['members_count'], ascending=False)\n",
    "         .reset_index(drop=True)\n",
    "         .query('cluster_rep_locus_tag.isin(@nodes_in_edges)')\n",
    "         .merge(rep_prokka_annotation_with_dfs_prediction_df, left_on='cluster_rep_locus_tag', right_on='locus_tag', how='left')\n",
    "         .drop('locus_tag', axis=1)\n",
    "         .merge((clusters_df\n",
    "                   .groupby('cluster_rep_locus_tag', as_index=False)\n",
    "                   .apply(lambda df: pd.DataFrame([(df.cluster_rep_locus_tag.values.tolist()[0], *max(tuple(df.value_counts('strand').to_dict().items()), key=lambda x: x[1]))], columns=['cluster_rep_locus_tag', 'strand', 'count']))\n",
    "                   .reset_index(drop=True)\n",
    "                   .sort_values('count', ascending=False)\n",
    "                   .rename(columns={'count' : 'most_frq_strnd_orint_count', 'strand': 'most_frq_strnd_orint'})),\n",
    "            on='cluster_rep_locus_tag',\n",
    "            how='left')\n",
    "         .rename(columns={'cluster_rep_locus_tag': 'id'}))\n",
    "nodes.to_csv('nodes_cache.csv', index=False, sep=';')\n",
    "nodes.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
